SSA/Projects/InstCombine/AliveHandwrittenExamples.lean

```diff
@@ -79,4 +257,439 @@ theorem alive_DivRemOfSelect (w : Nat) :
+        norm_num
+        conv =>
+          lhs
+          rw[Nat.mul_mod, Nat.mod_mod, ← Nat.mul_mod]
+        congr 1
+        ring
+
+end MulDivRem
+
+
+namespace AndOrXor
+/-
+Name: AndOrXor:2515   ((X^C1) >> C2)^C3 -> (X>>C2) ^ ((C1>>C2)^C3)
+%e1  = xor %x, C1
+%op0 = lshr %e1, C2
+%r   = xor %op0, C3
+  =>
+%o = lshr %x, C2 -- (X>>C2)
+%p = lshr(%C1,%C2)
+%q = xor %p, %C3 -- ((C1>>C2)^C3)
+%r = xor %o, %q
+-/
+
+open ComWrappers
+open LLVMTheory
+def AndOrXor2515_lhs (w : ℕ):
+  Com InstCombine.Op
+    [/- C1 -/ InstCombine.Ty.mkBitvec w,
+     /- C2 -/ InstCombine.Ty.mkBitvec w,
+     /- C3 -/ InstCombine.Ty.mkBitvec w,
+     /- %X -/ InstCombine.Ty.mkBitvec w] (InstCombine.Ty.mkBitvec w) :=
+  /- e1  = -/ Com.lete (xor w ⟨/-x-/ 0, by simp [Ctxt.snoc]⟩ ⟨ /-C1-/ 3, by simp [Ctxt.snoc]⟩) <|
+  /- op0 = -/ Com.lete (lshr w ⟨/-e1-/ 0, by simp [Ctxt.snoc]⟩ ⟨ /-C2-/ 3, by simp [Ctxt.snoc]⟩) <|
+  /- r   = -/ Com.lete (xor w ⟨/-op0-/ 0, by simp [Ctxt.snoc]⟩ ⟨ /-C3-/ 3, by simp [Ctxt.snoc]⟩) <|
+  Com.ret ⟨/-r-/0, by simp [Ctxt.snoc]⟩
+
+def AndOrXor2515_rhs (w : ℕ) :
+  Com InstCombine.Op
+    [/- C1 -/ InstCombine.Ty.mkBitvec w,
+     /- C2 -/ InstCombine.Ty.mkBitvec w,
+     /- C3 -/ InstCombine.Ty.mkBitvec w,
+     /- %X -/ InstCombine.Ty.mkBitvec w] (InstCombine.Ty.mkBitvec w) :=
+  /- o = -/ Com.lete (lshr w ⟨/-X-/ 1, by simp [Ctxt.snoc]⟩ ⟨/-C2-/2, by simp [Ctxt.snoc]⟩) <|
+  /- p = -/ Com.lete (lshr w ⟨/-C1-/ 4, by simp [Ctxt.snoc]⟩ ⟨/-C2-/3, by simp [Ctxt.snoc]⟩) <|
+  /- q = -/ Com.lete (xor w ⟨/-p-/0, by simp [Ctxt.snoc]⟩ ⟨/-C3-/3, by simp [Ctxt.snoc]⟩) <|
+  /- r = -/ Com.lete (xor w ⟨/-o-/2, by simp [Ctxt.snoc]⟩ ⟨/-q-/0, by simp [Ctxt.snoc]⟩) <|
+  Com.ret ⟨/-r-/0, by simp [Ctxt.snoc]⟩
+
+
+def alive_simplifyAndOrXor2515 (w : Nat) :
+  AndOrXor2515_lhs w ⊑ AndOrXor2515_rhs w := by
+  simp only [Com.Refinement]; intros Γv
+  simp only [AndOrXor2515_lhs, AndOrXor2515_rhs,
+    InstCombine.Ty.mkBitvec, const, shl, mul, Ctxt.get?, Var.zero_eq_last, add, icmp, ComWrappers.xor, lshr, select]
+  simp_peephole [MOp.sdiv, MOp.binary, MOp.BinaryOp.shl, MOp.shl, MOp.mul, MOp.and, MOp.or, MOp.xor, MOp.add] at Γv
+  intros x c3 c2 c1
+  simp only [InstCombine.Op.denote, pairBind, HVector.toPair, HVector.toTuple, Function.comp_apply, HVector.toTriple, bind_assoc]
+  rcases c1 with none | c1 <;>
+  rcases c2 with none | c2 <;>
+  rcases c3 with none | c3 <;>
+  rcases x with none | x <;> simp[Option.bind, Bind.bind]
+  . rcases (LLVM.lshr? (x ^^^ c1) c2) with none | irrelevant <;> simp
+  . -- TODO: develop automation to automatically do the case split. See also: prime motivation for simproc.
+    have hc2: (c2.toNat < w) ∨ ¬ (c2.toNat < w) := by tauto
+    rcases hc2 with lt | gt
+    . simp only [ge_iff_le, LLVM.lshr?_eq_some lt, BitVec.ushiftRight_eq,
+      BitVec.Refinement.some_some]
+      -- (x ^^^ c1) >>> BitVec.toNat c2 ^^^ c3 = c3 >>> BitVec.toNat c2 ^^^ (c1 >>> BitVec.toNat c2 ^^^ c3)
+      sorry -- I need bitwise reasoning here!
```

Comment:
@alexkeizer I was wondering how I finish this off with bitwise reasoning! I can't seem to recall the tactic incantation to do this. IIRC, we had tried the bitwise version once upon a time, but I am not sure how to use it.